suffuse/range.go

```diff
@@ -0,0 +1,79 @@
+package suffuse
+
+/** Go slices fail hard if the inputs don't make sense, but
+ *  slices are much more convenient to work with if they do the
+ *  sensible thing when you overshoot, e.g. slice[0:10] if there
+ *  are only 5 elements would give you the 5. Range superimposes
+ *  the scala-like forgiving slice logic on top of native slices.
+ */
+type Offset uint
+type Length uint
+
+type Range struct {
+  Offset
+  Length
+}
+
+/** All empty ranges are defined as 0:0. We don't want the
+ *  start value of an empty range to be considered meaningful
+ *  or stable under any operations.
+ */
+var EmptyRange Range = Range { Offset(0), Length(0) }
+
+func (x Range) IsEmpty()bool  { return x.LengthInt() == 0           }
+func (x Range) StartInt()int  { return int(x.Offset)                }
```

Comment:
Because Start is an Offset. I avoid returning types-which-aren't-types like int wherever possible, but I generally embed the name of the primitive type when I do so to highlight the fact.